src/utils/data_transformer/data_transformer.py

```diff
@@ -0,0 +1,252 @@
+) -> (CairoData, CairoData):
+    if len(values) < n:
+        raise ValueError(
+            f"Output {name} expected {n} values, {len(values)} values are available."
+        )
+
+    return values[:n], values[n:]
+
+
+UsedCairoType = TypeVar("UsedCairoType", bound=CairoType)
+PythonType = TypeVar("PythonType")
+
+
+@dataclass
+class TypeTransformer(Generic[UsedCairoType, PythonType]):
+    identifier_manager: IdentifierManager
+    resolve_type: Callable[[CairoType], "TypeTransformer"]
+
+    def from_python(
+        self, cairo_type: UsedCairoType, name: str, value: any
+    ) -> CairoData:
+        raise NotImplementedError()
+
+    def to_python(
+        self, cairo_type: UsedCairoType, name: str, values: CairoData
+    ) -> Tuple[PythonType, CairoData]:
+        raise NotImplementedError()
+
+
+class FeltTransformer(TypeTransformer[TypeFelt, int]):
+    def from_python(self, cairo_type, name, value):
+        if not isinstance(value, int):
+            raise TypeError(f"{name} should be int.")
+        return [value]
+
+    def to_python(self, cairo_type, name, values):
+        [val], rest = read_from_cairo_data(name, values, 1)
+        return val, rest
+
+
+class StructTransformer(TypeTransformer[TypeStruct, dict]):
+    def _definition(self, cairo_type: TypeStruct) -> StructDefinition:
+        definition = self.identifier_manager.get(
+            cairo_type.resolved_scope
+        ).identifier_definition
+
+        if not isinstance(definition, StructDefinition):
+            raise ValueError(
+                f"Invalid definition found for {cairo_type.resolved_scope}."
+            )
+
+        return definition
+
+    def from_python(self, cairo_type, name, value):
+        if not isinstance(value, dict):
+            raise TypeError(f"Expected {name} to be a dict.")
+
+        result = []
+
+        definition = self._definition(cairo_type)
+
+        for member_name, member in definition.members.items():
+            if member_name not in value:
+                raise TypeError(f"{name}[{member_name}] not provided.")
+
+            values = self.resolve_type(member.cairo_type).from_python(
+                member.cairo_type, f"{name}.{member_name}", value[member_name]
+            )
+            result.extend(values)
+
+        return result
+
+    def to_python(self, cairo_type, name, values) -> (dict, CairoData):
+        definition = self._definition(cairo_type)
+
+        result = {}
+        for member_name, member in definition.members.items():
+            transformed, values = self.resolve_type(member.cairo_type).to_python(
+                member.cairo_type, f"{name}.{member_name}", values
+            )
+            result[member_name] = transformed
+
+        return result, values
+
+
+class TupleTransformer(TypeTransformer[TypeTuple, tuple]):
+    def from_python(self, cairo_type, name, value):
+        values = [*value]
+        if len(values) != len(cairo_type.members):
+            raise ValueError(
```

Comment:
i think we should raise TypeErrors for those type transformers

Reply:
[Typ](https://docs.python.org/3/library/exceptions.html#TypeError)
>Passing arguments of the wrong type (e.g. passing a list when an int is expected) should result in a TypeError, but passing arguments with the wrong value (e.g. a number outside expected boundaries) should result in a ValueError.

Types match here, so it is a case for value error.